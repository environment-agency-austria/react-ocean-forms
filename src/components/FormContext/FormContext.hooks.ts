import { useContext } from 'react';

import { FormContext } from './FormContext';
import { IFieldValues, IFormContext } from './FormContext.types';

// tslint:disable-next-line:export-name naming-convention
export const useFormContext = <TFieldValues = IFieldValues>(): IFormContext<TFieldValues> => {
  const context = useContext(<React.Context<IFormContext<TFieldValues> | undefined>>FormContext);
  if (context === undefined) {
    throw new Error('[useFormContext]: Could not find form context. This component must be used inside a <Form> tag.');
  }

  return context;
};
