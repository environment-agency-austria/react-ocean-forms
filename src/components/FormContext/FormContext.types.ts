/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { TSTringFormatter } from '../../utils/stringFormatter';
import { TFieldValue } from '../withField';
import { IValidationArgs, IValidationState } from '../withValidation';

export type TFormEventListener = ((name: string, event: string, args?: unknown) => void);

/**
 * Interface describing field states
 */
export interface IFieldState {
  /**
   * Label of the field
   */
  label: string;
  /**
   * True if the field is actually a FieldGroup
   */
  isGroup?: boolean;
  /**
   * Triggers the validation of this field
   * @param args Validation args
   */
  validate(args?: Partial<IValidationArgs>): Promise<IValidationState>;
  /**
   * Returns the current value of the field
   */
  getValue(): TFieldValue;
  /**
   * Resets the field to its initial state
   */
  reset(): void;
  /**
   * Updates the validation state of the field
   * @param state New validation state
   */
  updateValidation(state: Partial<IValidationState>): void;
}

/**
 * Type describing a collection of field values
 */
export type IFieldValues = {
  [prop: string]: TFieldValue | IFieldValues;
};

/**
 * Base interface for the form context
 */
export interface IBaseFormContext<TFieldValues = IFieldValues> {
  /**
   * Optional field prefix
   */
  fieldPrefix: string | null;

  /**
   * True, if the form is currently busy
   */
  busy: boolean;

  /**
   * Registers a field to the form context
   * @param name Unique field name
   * @param state Field state
   */
  registerField(name: string, state: IFieldState): void;
  /**
   * Unregisters the field from the form context
   * @param name Unique field name
   */
  unregisterField(name: string): void;

  /**
   * Notifies the form context about a field event
   * @param name Unique field name
   * @param event Event name
   * @param args Optional event args
   */
  notifyFieldEvent(name: string, event: string, args?: unknown): void;

  /**
   * Registers a listener to all field events inside
   * this form context
   * @param name Unique listener name
   * @param callback Callback method
   */
  registerListener(name: string, callback: TFormEventListener): void;
  /**
   * Unregisters a listener from the form context
   * @param name Unique listener name
   */
  unregisterListener(name: string): void;

  /**
   * Returns the field state of the given field
   * @param name Unique field name
   */
  getFieldState(name: string): IFieldState;
  /**
   * Returns all field values
   */
  getValues(): TFieldValues;

  /**
   * Submits the form
   * @param submitArgs Optional submit args
   */
  submit(submitArgs?: unknown): Promise<void>;
}

/**
 * Interface describing the form context
 */
export interface IFormContext<TFieldValues = IFieldValues> extends IBaseFormContext<TFieldValues> {
  /**
   * Contains the default values of the form. Those values will be
   * put into the according fields when the form initializes.
   */
  defaultValues: Partial<TFieldValues>;
  /**
   * Contains the values of the form. Changing this property will
   * update all Field values, overwriting their default values but also
   * any value the user put in.
   */
  values?: Partial<TFieldValues>;
  /**
   * If set to true the form will trigger asynchronous validation on
   * Fields whenever they change (e.g. on key press). Default behaviour
   * is that the fields will only async validate when they loose focus.
   * Can be overriden per field.
   */
  asyncValidateOnChange: boolean;
  /**
   * Configures the wait time before the async validators will be called.
   * Per default the form will call the async validators only 400ms after
   * the last value change. Can be overriden per field.
   */
  asyncValidationWait: number;
  /**
   * Optional string formatter method
   */
  stringFormatter: TSTringFormatter;
  /**
   * If set to true the form will disable all Fields and FormButtons.
   */
  disabled: boolean;
  /**
   * If set to true, all the fields will display only text instead of an
   * input element. This is useful to re-use Fields in a check page.
   */
  plaintext: boolean;
}
