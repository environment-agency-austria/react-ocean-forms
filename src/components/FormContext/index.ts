export * from './FormContext';
export * from './FormContext.types';
export * from './FormContext.hooks';
