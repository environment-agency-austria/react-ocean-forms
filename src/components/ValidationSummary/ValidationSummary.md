### Examples

#### Showcase

Showcase of the validation summary. Click on submit to display the summary.

```jsx
import { Form, Input, ValidationSummary, validators } from 'react-ocean-forms';

function Example({ logMessage }) {
  return (
    <Form className="demo">
      <ValidationSummary id="summary" />

      <Input name="input1" label="Example input 1" validators={[validators.required]} />

      <Input name="input2" label="Example input 2" validators={[validators.required]} />

      <button type="submit">Submit</button>
      <button type="reset">Reset</button>
    </Form>
  );
}

<Example />;
```
