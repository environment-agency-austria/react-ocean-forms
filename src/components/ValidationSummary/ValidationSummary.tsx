/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as React from 'react';

import { toArray } from '../../utils';
import { IFieldErrorObject } from '../../validators';
import { IFormContextProps, withForm } from '../withForm';
import { IValidationState } from '../withValidation';

interface IValidationSummaryProps extends IFormContextProps {
  id: string;
  title: string;
  disableFocusOnSubmit: boolean;
  renderFieldError?(id: string, fieldName: string, errors: React.ReactNode, linkCallback: React.MouseEventHandler): JSX.Element;
  render?(children: JSX.Element): JSX.Element;
}

interface IValidationEventArgs extends IValidationState {
  label: string;
}

interface IInvalidField {
  id: string;
  name: string;
  error: IFieldErrorObject | IFieldErrorObject[];
}

interface IFieldContainer {
  [s: string]: IValidationEventArgs;
}

interface IValidationSummaryState {
  fields: IFieldContainer;
}

/**
 * Component for displaying a summary of all
 * validation errors of the form this component
 * lives in.
 */
export class BaseValidationSummary extends React.Component<IValidationSummaryProps, IValidationSummaryState> {
  public static displayName: string = 'ValidationSummary';

  // tslint:disable-next-line:typedef
  public static defaultProps = {
    title: 'ojs_form_validationSummaryHeader',
    disableFocusOnSubmit: false,
  };

  private headerRef: React.RefObject<HTMLHeadingElement>;

  constructor(props: IValidationSummaryProps) {
    super(props);

    this.headerRef = React.createRef();

    this.notify = this.notify.bind(this);
    this.scrollIntoView = this.scrollIntoView.bind(this);

    const { context: { registerListener }, id } = props;
    registerListener(id, this.notify);

    this.state = {
      fields: {},
    };
  }

  /**
   * Renders a field error
   * @param id Unique id
   * @param fieldName Name of the field
   * @param errors Field errors
   * @param linkCallback Callback for the error click
   */
  private static renderError(id: string, fieldName: string, errors: JSX.Element[], linkCallback: React.MouseEventHandler): JSX.Element {
    return (
      <li key={id}>
        <a href={`#${id}`} onClick={linkCallback}>
          {fieldName}
          :
          {' '}
          {errors}
        </a>
      </li>
    );
  }

  /**
   * Unregisters the field from the form
   */
  public componentWillUnmount(): void {
    const { context, id } = this.props;
    context.unregisterListener(id);
  }

  /**
   * Gets called when a validation state changes
   * @param name Field name
   * @param event Event name
   * @param state Field state
   */
  private notify(name: string, event: string, state: IValidationEventArgs): void {
    if (event === 'validation') {
      this.setState(oldState => ({
        fields: {
          ...oldState.fields,
          [name]: state,
        },
      }));
    } else if (event === 'submit-invalid') {
      this.scrollIntoView();
    }
  }

  private scrollIntoView(): void {
    const { disableFocusOnSubmit } = this.props;
    if (this.headerRef.current && !disableFocusOnSubmit) {
      this.headerRef.current.scrollIntoView({ block: 'start', behavior: 'smooth' });
    }
  }

  /**
   * Renders a list of all errors
   * @param errors Error array
   */
  private renderErrors(errors: IInvalidField[]): JSX.Element[] {
    const {
      context: {
        stringFormatter,
      },
    } = this.props;

    return errors.map((error: IInvalidField) => {
      const errorArray = toArray(error.error);
      const fieldErrors = errorArray.map((item) => {
        const errorString = stringFormatter(item.message_id, item.params);

        return (
          <span key={`${error.id}_${item.message_id}`}>{errorString}</span>
        );
      });

      // Focuses the invalid field on click on the error message
      const linkCallback = (event: React.MouseEvent<Element>): void => {
        event.preventDefault();
        const input = document.getElementById(error.id);
        if (input) {
          input.focus();
        }
      };

      const fieldName = stringFormatter(error.name);

      // Use an optional render prop if the user
      // wants to customize the output of the error
      const { renderFieldError } = this.props;
      if (renderFieldError !== undefined) {
        return renderFieldError(
          error.id,
          fieldName,
          fieldErrors,
          linkCallback,
        );
      }

      return BaseValidationSummary.renderError(
        error.id,
        fieldName,
        fieldErrors,
        linkCallback,
      );
    });
  }

  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element | null {
    const {
      id,
      title,
      context: {
        stringFormatter,
      },
    } = this.props;

    const { fields } = this.state;

    // Fetch the current errors out of the form context
    const fieldArray = Object.entries(fields);
    const errors: IInvalidField[] = [];

    fieldArray.forEach(([name, state]) => {
      if (state.valid === false && state.error !== null) {
        errors.push({
          id: name,
          name: state.label,
          error: state.error,
        });
      }
    });

    // Don't render anything if there are no errors
    if (errors.length === 0) { return null; }

    const renderedErrors = this.renderErrors(errors);
    const titleString = stringFormatter(title);

    const summary = (
      <div id={id} className="validation-summary">
        <h4 className="alert-heading" ref={this.headerRef}>{titleString}</h4>
        <ul>
          {renderedErrors}
        </ul>
      </div>
    );

    // Use an optional render prop if the user
    // wants to customize the look and feel of
    // the validation summary
    const { render: renderProp } = this.props;
    if (renderProp !== undefined) {
      return renderProp(summary);
    }

    return summary;
  }
}

export const ValidationSummary = withForm(BaseValidationSummary);
