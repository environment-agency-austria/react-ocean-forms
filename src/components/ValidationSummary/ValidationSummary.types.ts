/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Properties for the ValidationSummary
 */
export interface IValidationSummaryProps {
  /**
   * Id of this input. Will be used as the unique identifier of the div.
   * **Must be unique form wide!**
   */
  id: string;
  /**
   * Message that will be used as the title, wrapped in a h4.
   */
  title?: string;
  /**
   * If set to true the validation summary will stop automatically scrolling
   * to itself when the user clicks on a submit button and the form is invalid.
   * @default false
   */
  disableFocusOnSubmit?: boolean;
  /**
   * Optional function that can be used to customize the output of each error.
   * @param id Id of the failing field
   * @param fieldName Display name of the failing field
   * @param errors Object / Array of objects with the rendered error messages
   * @param linkCallback Function that should be called on click, will focus on the input element.
   */
  renderFieldError?(id: string, fieldName: string, errors: React.ReactNode, linkCallback: React.MouseEventHandler): JSX.Element;
  /**
   * Optional function that can be used to customize the ValidationSummary rendering.
   * Note: This function will only be called when there are errors to be displayed.
   * @param children Validation summary content (header + rendered field errors)
   */
  render?(children: JSX.Element): JSX.Element;
}
