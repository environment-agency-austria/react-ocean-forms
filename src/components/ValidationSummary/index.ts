export * from './ValidationSummary';
export * from './ValidationSummary.types';
