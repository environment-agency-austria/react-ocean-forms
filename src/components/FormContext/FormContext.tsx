/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';
import { IFormContext } from './FormContext.types';

/**
 * Context for the communication between the form
 * and its fields.
 */
export const FormContext = React.createContext<IFormContext | undefined>(undefined);
