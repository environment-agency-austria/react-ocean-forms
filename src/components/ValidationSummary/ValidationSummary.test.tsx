import * as React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';

import { mockEvent } from '../../test-utils/enzymeEventUtils';
import { createMockFormContext } from '../../test-utils/enzymeFormContext';
import { TFormEventListener } from '../FormContext';
import { BaseValidationSummary } from './ValidationSummary';
import { IValidationSummaryProps } from './ValidationSummary.types';

describe('<ValidationSummary />', () => {
  let listenerCallback: TFormEventListener;
  const registerCallback = (name: string, state: TFormEventListener): void => { listenerCallback = state; };
  const formContext = createMockFormContext(registerCallback);

  const summaryId = 'unitSummary';
  const setup = (props?: Partial<IValidationSummaryProps>): ShallowWrapper => shallow((
    <BaseValidationSummary
      id={summaryId}
      context={formContext}
      {...props}
    />
  ));
  const wrapper = setup();

  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should register itself in the form context', () => {
    expect(formContext.registerListener).toHaveBeenCalledWith(
      summaryId,
      expect.any(Function),
    );
  });

  const errorFieldName = 'unitField';
  const errorFieldLabel = 'Unit field';
  const mockError = {
    message_id: 'test',
    params: {},
  };

  it('should call the render props correctly', () => {
    listenerCallback(
      errorFieldName,
      'validation',
      {
        label: errorFieldLabel,
        valid: false,
        error: mockError,
      },
    );
    wrapper.update();

    const mockRender = jest.fn().mockReturnValue(null);
    const mockErrorRender = jest.fn().mockReturnValue(null);
    wrapper.setProps({ render: mockRender, renderFieldError: mockErrorRender });

    expect(mockRender).toHaveBeenCalled();
    expect(mockRender).toMatchSnapshot();

    expect(mockErrorRender).toHaveBeenCalledWith(
      errorFieldName,
      errorFieldLabel,
      expect.any(Array),
      expect.any(Function),
    );
  });

  it('should render when it receives new validation states', () => {
    wrapper.setProps({ render: undefined, renderFieldError: undefined });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  it('should ignore valid fields', () => {
    listenerCallback(
      'validField',
      'validation',
      {
        label: errorFieldLabel,
        valid: true,
        error: null,
      },
    );
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  it('should focus on the field when you click on the error message', () => {
    const inputNode = document.createElement('input');
    inputNode.type = 'text';
    inputNode.id = errorFieldName;

    document.body.appendChild(inputNode);
    const spy = jest.spyOn(inputNode, 'focus');

    const errorLink = wrapper.find('ul li a');
    errorLink.simulate('click', mockEvent());

    expect(spy).toHaveBeenCalled();
    document.body.removeChild(inputNode);

    errorLink.simulate('click', mockEvent());
  });

  it('should scroll into view when called by the context', () => {
    // TODO: Can someone test this?
    listenerCallback('_form', 'submit-invalid');
  });

  it('should unregister on unmount', () => {
    wrapper.unmount();
    expect(formContext.unregisterListener).toHaveBeenCalledWith(summaryId);
  });
});
