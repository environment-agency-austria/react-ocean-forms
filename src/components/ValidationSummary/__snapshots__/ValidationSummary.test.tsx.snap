// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ValidationSummary /> render props render prop should be called with the basic summary render if the form is invalid 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <div
    className="validation-summary"
    id="unitSummary"
  >
    <h4
      className="alert-heading"
    >
      Errors
    </h4>
    <ul>
      <li>
        <a
          href="#unitField"
          onClick={[Function]}
        >
          Unit field
          :
           
          <span>
            test
          </span>
        </a>
      </li>
    </ul>
  </div>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h4
          className="alert-heading"
        >
          Errors
        </h4>,
        <ul>
          <li>
            <a
              href="#unitField"
              onClick={[Function]}
            >
              Unit field
              :
               
              <span>
                test
              </span>
            </a>
          </li>
        </ul>,
      ],
      "className": "validation-summary",
      "id": "unitSummary",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Errors",
          "className": "alert-heading",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": "Errors",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <li>
              <a
                href="#unitField"
                onClick={[Function]}
              >
                Unit field
                :
                 
                <span>
                  test
                </span>
              </a>
            </li>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "unitField",
            "nodeType": "host",
            "props": Object {
              "children": <a
                href="#unitField"
                onClick={[Function]}
              >
                Unit field
                :
                 
                <span>
                  test
                </span>
              </a>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Unit field",
                  ":",
                  " ",
                  Array [
                    <span>
                      test
                    </span>,
                  ],
                ],
                "href": "#unitField",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                "Unit field",
                ":",
                " ",
                Object {
                  "instance": null,
                  "key": "unitField_test",
                  "nodeType": "host",
                  "props": Object {
                    "children": "test",
                  },
                  "ref": null,
                  "rendered": "test",
                  "type": "span",
                },
              ],
              "type": "a",
            },
            "type": "li",
          },
        ],
        "type": "ul",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h4
            className="alert-heading"
          >
            Errors
          </h4>,
          <ul>
            <li>
              <a
                href="#unitField"
                onClick={[Function]}
              >
                Unit field
                :
                 
                <span>
                  test
                </span>
              </a>
            </li>
          </ul>,
        ],
        "className": "validation-summary",
        "id": "unitSummary",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Errors",
            "className": "alert-heading",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": "Errors",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <li>
                <a
                  href="#unitField"
                  onClick={[Function]}
                >
                  Unit field
                  :
                   
                  <span>
                    test
                  </span>
                </a>
              </li>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "unitField",
              "nodeType": "host",
              "props": Object {
                "children": <a
                  href="#unitField"
                  onClick={[Function]}
                >
                  Unit field
                  :
                   
                  <span>
                    test
                  </span>
                </a>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Unit field",
                    ":",
                    " ",
                    Array [
                      <span>
                        test
                      </span>,
                    ],
                  ],
                  "href": "#unitField",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": Array [
                  "Unit field",
                  ":",
                  " ",
                  Object {
                    "instance": null,
                    "key": "unitField_test",
                    "nodeType": "host",
                    "props": Object {
                      "children": "test",
                    },
                    "ref": null,
                    "rendered": "test",
                    "type": "span",
                  },
                ],
                "type": "a",
              },
              "type": "li",
            },
          ],
          "type": "ul",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`<ValidationSummary /> should render without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ValidationSummary
    context={
      Object {
        "asyncValidateOnChange": false,
        "asyncValidationWait": 400,
        "busy": false,
        "defaultValues": Object {},
        "disabled": false,
        "fieldPrefix": null,
        "getFieldState": [MockFunction],
        "getValues": [MockFunction],
        "notifyFieldEvent": [MockFunction],
        "plaintext": false,
        "registerField": [MockFunction],
        "registerListener": [MockFunction] {
          "calls": Array [
            Array [
              "unitSummary",
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "stringFormatter": [MockFunction],
        "submit": [MockFunction],
        "unregisterField": [MockFunction],
        "unregisterListener": [MockFunction],
      }
    }
    disableFocusOnSubmit={false}
    id="unitSummary"
    title="ojs_form_validationSummaryHeader"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
